databuilder/dummy_data/generator.py

```diff
@@ -1,20 +1,142 @@
-from databuilder.orm_utils import orm_classes_from_qm_tables
+import random
+from datetime import date, timedelta
+
+from sqlalchemy.orm import declarative_base
+
+from databuilder.dummy_data.query_info import QueryInfo
+from databuilder.orm_utils import orm_class_from_schema
 from databuilder.query_engines.in_memory import InMemoryQueryEngine
 from databuilder.query_engines.in_memory_database import InMemoryDatabase
-from databuilder.query_model import get_table_nodes
 
 
 class DummyDataGenerator:
+    # TODO: Make these user configurable. We're deliberately not doing this right now
+    # because we want to keep the API surface to zero in the early stages while we
+    # iterate on things.
+    population_size = 5000
+    random_seed = "BwRV3spP"
+
     def __init__(self, variable_definitions):
         self.variable_definitions = variable_definitions
-        self.tables = get_table_nodes(*variable_definitions.values())
-        self.orm_classes = orm_classes_from_qm_tables(self.tables)
+
+        # TODO: I dislike using today's date as part of the data generation because it
+        # makes the results non-deterministic. However until we're able to infer a
+        # suitable time range by inspecting the query, this will have to do.
+        self.today = date.today()
+        self.rnd = random.Random()
+
+        # Create ORM classes for each of the tables used in the dataset definition
+        self.query_info = QueryInfo.from_variable_definitions(self.variable_definitions)
+        Base = declarative_base()
+        self.orm_classes = {
+            table_info.name: orm_class_from_schema(
+                Base,
+                table_info.name,
+                table_info,
+                table_info.has_one_row_per_patient,
+            )
+            for table_info in self.query_info.tables.values()
+        }
 
     def get_data(self):
-        return [orm_class(patient_id=1) for orm_class in self.orm_classes]
+        for patient_id in range(1, self.population_size + 1):
+            yield from self.get_patient_data(patient_id)
 
     def get_results(self):
         database = InMemoryDatabase()
         database.setup(self.get_data())
         engine = InMemoryQueryEngine(database)
         return engine.get_results(self.variable_definitions)
+
+    def get_patient_data(self, patient_id):
```

Comment:
Should this be a method on a different class?  I think this class is carrying too much state -- and it's responsible both for generating all dummy data, and for generating data for one patient.

Reply:
I actually originally had it like that and then I felt like there were too many classes floating around and it didn't help much. But that was before the extra complexity I added with the batching/filtering and lazy generation. So I think you're right now. I'll split them back out.

Reply:
Doesn't have to happen before half term!